universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java

```diff
@@ -193,4 +221,59 @@ private long getNextBlockId(ProgramDefinition programDefinition) {
             .orElseGet(() -> 0L)
         + 1;
   }
+
+  /**
+   * Update all {@link QuestionDefinition}s in the ProgramDefinition with appropriate versions from
+   * the {@link QuestionService}.
+   */
+  private ProgramDefinition syncQuestions(ProgramDefinition programDefinition) {
+    return questionService
+        .getReadOnlyQuestionService()
+        .thenApplyAsync(
+            (roQuestionService) -> {
+              return syncProgramDefinitionQuestions(programDefinition, roQuestionService);
+            },
+            httpExecutionContext.current())
+        .toCompletableFuture()
+        .join();
+  }
```

Comment:
Issue (blocking): this is called by `getProgramDefinitionAsync`, which the applicant service consumes, but blocks on database I/O in `getReadOnlyQuestionService`.

Suggestion: make this method async and let synchronous code use `join` on the return value.

Reply:
done